process.cpuUsage() - implementation, doc, tests

pmuellr opened the pull request:
<!--
Thank you for submitting a pull request to Node.js. Before you submit, please
review below requirements and walk through the checklist. You can 'tick'
a box by using the letter "x": [x].

Run the test suite by invoking: `make -j4 lint test` on linux or
`vcbuild test nosign` on Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. Finally – if possible – a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist

<!-- remove lines that do not apply to you -->

- [x] tests and code linting passes (`make -j4 lint test`)
- [x] a test and/or benchmark is included
- [x] documentation is changed or added
- [x] the commit message follows commit guidelines

##### current planned commit message

process: add `cpuUsage()`

The `cpuUsage()` function will return user and system CPU usage timings for the
current process.

Drive-by fix to process.hrtime() doc to align with preferred non-usage of "you"
in function description.


##### Affected core subsystem(s)

`process` module


##### Description of change

add the function `cpuUsage()` to `process`

pmuellr commented:
See discussion from original PR https://github.com/nodejs/node/pull/5565

Fishrock123 commented on test/parallel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,47 @@
+var assert = require('assert');
+
+// the default behavior, return { user: [int, int], system: [int, int]}
+var result = process.cpuUsage();
```
`const` please :)

addaleax commented:
Hm, is there any concrete advantage to splitting up the 64-bit numbers into two lanes and reassembling them in JS? Like, wouldn’t using a `Float64Array` + `double`s be a bit easier and have the same results?

eljefedelrodeodeljefe commented on doc/api/process.markdown:
```diff
@@ -495,6 +495,29 @@ An example of the possible output looks like:
+while (Date.now() - now < 500);
+
+console.log(process.cpuUsage(startUsage));
+// { user: [ 0, 509627000 ], system: [ 0, 9993000 ] }
```
nit: `const`

eljefedelrodeodeljefe commented on doc/api/process.markdown:
```diff
@@ -495,6 +495,29 @@ An example of the possible output looks like:
+system code respectively, and may end up being greater than actual elapsed time
+if multiple cpu cores are performing work for this process.
+
+You may pass in the result of a previous call to `process.cpuUsage()` to get a
```
nit: avoid `you` reference, maybe: `process.cpuUsage() can optionally take a previous call as argument to get a diff reading`

pmuellr replied:
Sounds good.  Note, I copied off of `hrtime()` which has the same usage of 'You ...', because that's what I copied from.  Happy to do a drive-by fix to that as well ...

eljefedelrodeodeljefe replied:
That'd be awesome. Hope no-one objects; it would save me a PR, since at some point, we wanted to review the usage those. Thx.

eljefedelrodeodeljefe commented on doc/api/process.markdown:
```diff
@@ -495,6 +495,29 @@ An example of the possible output looks like:
+with properties `user` and `system`, whose values are arrays of integers as
+`[seconds, nanoseconds]` tuples. These values measure time spent in user and
+system code respectively, and may end up being greater than actual elapsed time
+if multiple cpu cores are performing work for this process.
```
nit: `CPU`

eljefedelrodeodeljefe commented:
Awesome. LGTM with nits.

pmuellr commented:
@addaleax I copied the style of `hrtime()` for `cpuUsage()`.  I think it may have been possible to combine all the numeric bits into a single large int, because the cpu usage values are to the microsecond precision, instead of nanosecond (like `hrtime()`).  Figured it would be better to copy the style instead of doing something radically different.

addaleax commented:
@pmuellr Sounds reasonable – was just wondering because it seemed a little odd :)

mhdawson commented:
I'd suggest a CI run so that we can see if uv_getrusage is fully supported across all platforms.  Otherwise we might need to add info in that respect to the docs.  I can't tell if you can do that so I launched one for you:

https://ci.nodejs.org/job/node-test-pull-request/2242/

eljefedelrodeodeljefe commented:
Should be fine. On Windows only those two fields this PR is using are populated. The unices share the same function, but don't populate same results for some fields. nit: The inline comment could be rephrased, omitting the part that `uv_getrusage` is exposed.

Fishrock123 commented:
cc @mscdex who appears to have primarily reviewed the previous PR.

mscdex commented:
I didn't really review the last PR, I was mostly just seeing if there was interest in having a separate `.metrics` or similarly-named namespace for memory, cpu, other resource usage.

Just briefly looking over this PR, I'm wondering if it might just be better to drop the extra `* 1000` used to keep the same units as `hrtime()`. I think it would help in the case if a lot of CPU time is used by the process, where it would reach `Number.MAX_SAFE_INTEGER` much quicker than without the extra multiplication. `hrtime()` uses nanoseconds because it actually has that kind of resolution, but these CPU usage values do not.

jasnell commented on lib/internal/process.js:
```diff
@@ -22,6 +23,65 @@ const assert = process.assert = function(x, msg) {
+
+  process.cpuUsage = function cpuUsage(prevValue) {
+    const err = _cpuUsage(cpuValues);
+    if (err !== 0) throw new Error('unable to obtain cpu usage time');
```
Just a style nit...

```js
if (_cpuUsage(cpuValues) !== 0)
  throw new Error('unable to obtain cpu usage time');
```

jasnell commented:
I'd be +1 on not having the additional `* 1000`

jasnell commented on test/pummel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,30 @@
+require('../common');
+var assert = require('assert');
+
+var start = process.cpuUsage();
```
definitely use `const` wherever possible :-)

jasnell commented:
Generally LGTM with a few nits.

Fishrock123 commented:
@pmuellr I'd probably just make the commit message something like
```
process: add `cpuUsage()`
```

pmuellr commented:
@mhdawson that build is green, so guess it passes muster?

* https://ci.nodejs.org/job/node-test-pull-request/2243/

pmuellr commented:
@eljefedelrodeodeljefe 

> The inline comment could be rephrased, omitting the part that uv_getrusage is exposed.

My usual excuse here - I copied from `Hrtime()` just above it :-)

* https://github.com/pmuellr/node/blob/c9aae2dfad942b60a6d887830319cdd20ec47897/src/node.cc#L2200

pmuellr commented:
@mscdex 

> I'm wondering if it might just be better to drop the extra * 1000 used to keep the same units as hrtime(). I think it would help in the case if a lot of CPU time is used by the process, where it would reach Number.MAX_SAFE_INTEGER much quicker than without the extra multiplication. hrtime() uses nanoseconds because it actually has that kind of resolution, but these CPU usage values do not.

I think the argument comes down to aligning w/`hrtime()`'s nano-second resolution, to presumably make life easier for users, or to stay true to the `rusage()` precision or just optimize things a bit.

I'm siding with making life easier for users.

mscdex commented:
@pmuellr What does aligning with `hrtime()` output really gain us though? Are you expecting users to pass values from `cpuUsage()` to `hrtime()` or vice-versa?

pmuellr commented:
I've added the proposed commit message in the header of this PR.  Currently:

----

process: add `cpuUsage()`

The `cpuUsage()` function will return user and system CPU usage timings for the
current process.

Drive-by fix to process.hrtime() doc to align with preferred non-usage of "you"
in function description.

pmuellr commented:
@mscdex ah, yes, forgot to mention that

One expected uses of `cpuUsage()` would be to get the % of cpu this process is currently using.  You can do that by dividing a `cpuUsage()` diff from a `hrtime()` diff.  Example:

```
'use strict'

var startTime  = process.hrtime()
var startUsage = process.cpuUsage()

// spin the CPU for 500 milliseconds
var now = Date.now()
while (Date.now() - now < 500)

var elapTime = process.hrtime(startTime)
var elapUsage = process.cpuUsage(startUsage)

var elapTimeMS = secNSec2ms(elapTime)
var elapUserMS = secNSec2ms(elapUsage.user)
var elapSystMS = secNSec2ms(elapUsage.system)
var cpuPercent = Math.round(100 * (elapUserMS + elapSystMS) / elapTimeMS)

console.log('elapsed time ms:  ', elapTimeMS)
console.log('elapsed user ms:  ', elapUserMS)
console.log('elapsed system ms:', elapSystMS)
console.log('cpu percent:      ', cpuPercent)

function secNSec2ms (secNSec) {
  return secNSec[0] * 1000 + secNSec[1] / 1000000
}
```

Whenever you're dealing with `hrtime()`, it seems you will likely be wanting a single number to deal with, which you can determine the precision of yourself.  In this case, it converted the tuples into millisecond values.  Reused the same conversion function for both `hrtime()` and `cpuUsage()` values.

mscdex commented:
@pmuellr Right, but what I meant was you're not going to be doing `process.cpuUsage(startTime)` or `process.hrtime(startUsage)`.

I think I am still -1 on artificially adjusting the cpu usage values.

pmuellr commented:
@mscdex yes, correct, you'd never mix/match values from `cpuUsage()` and `hrtime()` to get diff values.

Still think there's value in having a shared model for these, so you can have a single (tuple) -> value conversion fn that works with both.

But it's not a deal-breaker for me or anything.

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,65 @@ const assert = process.assert = function(x, msg) {
+
+    // the second element of user/system comes back from libuv as
+    // microseconds, but we convert to nanoseconds here (* 1000) to
+    // be consistent with hrtime()
```
Comments should be capitalized and punctuated.

trevnorris commented on doc/api/process.markdown:
```diff
@@ -495,6 +495,29 @@ An example of the possible output looks like:
 
 If `process.connected` is false, it is no longer possible to send messages.
 
+## process.cpuUsage()
```
code allows passing `prevValue`, but is not documented.

trevnorris commented on src/node.cc:
```diff
@@ -2216,6 +2216,35 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+  fields[7] = (rusage.ru_stime.tv_usec) & 0xffffffff;
+
+  // return value will be return code from the libuv call
+  args.GetReturnValue().Set(err);
```
move this up just under the `uv_getrusage()` call. no sense in filling the fields with potential garbage if the operation failed.

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,65 @@ const assert = process.assert = function(x, msg) {
+      ],
+      system: [
+        cpuValues[4] * 0x100000000 + cpuValues[5],
+        (cpuValues[6] * 0x100000000 + cpuValues[7]) * 1000
```
Personally I'd go with `seconds.microseconds`. The double value has high enough resolution, and takes less time to operate on. So I'd switch the array to a `Float64Array` and assign `rusage.ru_utime.tv_sec + rusage.ru_utime.tv_usec / 1e6`. This type of operation doesn't require nanosecond precision.

pmuellr commented:
A few other notes on precision:

* on Windows, we only get millisecond precision anyway - see: https://github.com/libuv/libuv/blob/master/src/win/util.c#L1118-L1151

* if we go with microseconds (or even round to millis), then the final value can prolly be expressed as a single number, instead of tuple.  Back of the envelop calculations indicate we would roll over the number value at a process that used ~100 CPU years - because remember the time base here is from the start of the process, not epoch (like `hrtime()`)

Let me do the right work on when a rollover would occur with a single number value ... as having a single number, at either micro/millisecond precision seems better than the tuple story ...

eljefedelrodeodeljefe commented:
For the sake of documenting this, still -1 on having this in millis. The `hrtime()` API is already convoluted (though necessarily), but at least known to users. With unix `getrusage()` we are facing a **very** inconsistent unix API (even w/o counting in NT) and don't just need to inherit the inconsistency. From a data point of view I agree that nanos don't make too much sense. Working on this and frequently converting with google is somewhat bad though. Sorry. Also @pmuellr sorry for the copy-paste-nitpicks :D

Also it is not unrealistic that we gain nano seconds for cpu cycles when some tracing implementation would come around http://www.brendangregg.com/DTrace/cputimes

jasnell commented:
ping @pmuellr ... just checking in on this.

also, this looks to be covering the same ground as the older https://github.com/nodejs/node/pull/1568

eljefedelrodeodeljefe commented:
@jasnell they overlap, however I would say the difference could be the the nsec-usec conversion to gives `.cpuUsage()` the scope. So having both seems okay.

pmuellr commented:
I just added a commit with:

* change to use Float64 values of the user/system times in microsecond units, instead of second / nanosecond tuples.
* addressed current set of comments in this PR

@trevnorris [indicated a preference of second based units](https://github.com/nodejs/node/pull/6157#discussion-diff-59412028), but seemed a little more natural to return microsecond units as that is the current precision from libuv.  Diff values would then be "integral" instead of "fractional".  I can go either way on this.

I'm a bit torn on the issue of the overlap with https://github.com/nodejs/node/pull/1568 . It's an obvious overlap, but presumably if all you need is the CPU usage stats, this provides an easier to deal with value; a single number vs two numbers you'd need to do math on.

pmuellr commented:
Not sure if I should rebase | rebase-n-squash the current commits, based on the conflict.  Wait until the current commit is reviewed, then do that?  And shall I squash the commits if I'm going to rebase, or attempt to keep the individual commits (may be beyond my git powers, heh).

eljefedelrodeodeljefe commented:
Squashing is probably not necessary b/c GitHub squash. You need to make the doc change though, since the file has been renamed to `.md`

pmuellr commented:
rebased to master w/no squashing

trevnorris commented on doc/api/process.md:
```diff
@@ -793,15 +816,16 @@ if (process.getuid) {
-a diff reading, useful for benchmarks and measuring intervals:
+The result of a previous call to `process.hrtime()` can be passed as the
+argument to the function, to get a diff reading, which can be useful for
+benchmarks and measuring intervals:
```
This doc change should be in a different commit. It's unrelated to the `cpuUsage()` change. If the later needs to be reverted don't want to loose the former.

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,46 @@ const assert = process.assert = function(x, msg) {
+    };
+
+    // If a previous value was passed in, ensure it has the correct shape.
+    if (!Number.isFinite(prevValue.user)) {
```
May want to change this to `Number.isInteger()`, and to be even safer may use the following:
```js
if (!Number.isInteger(prevValue.user) ||
    prevValue.user > Number.MAX_SAFE_INTEGER ||
    prevValue.user < 0) {
  throw new TypeError('previous value "user" is not a safe integer');
}
```
Also noticed how I changed `Error` below to `TypeError`.

pmuellr replied:
* `Number.isInteger()` - that's checking an implementation detail.  If for some reason we get nanosecond resolution later, with a different implementation API (other than libuv's rusage fn), then that constraint no longer holds.

* `> Number.MAX_SAFE_INTEGER` check - One of the advantages of using floats for these values is so that we don't have to worry about such things.  The # of microseconds that fit in MAX_SAFE_INTEGER is ~250 years worth, making it just on the cusp of ridiculous that we'd overflow that, but "who would ever need more than 640KB of RAM?" :-)

trevnorris replied:
> One of the advantages of using floats for these values is so that we don't have to worry about such things.

Reason for using floats is so we have access to the 2^53 integer bits in a double. We're doing sanity checks to make sure users don't pass in unexpected values. So we check the limits. Is it an integer? Is it less than 2^53? Is it greater than zero? If those conditions are not met then it was called incorrectly.

trevnorris commented on src/node.cc:
```diff
@@ -2220,6 +2221,35 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+
+  // Set the Float64Array elements to be user / system values in microseconds.
+  fields[0] = 1.0L * rusage.ru_utime.tv_sec * MICROS_PER_SEC +
+    rusage.ru_utime.tv_usec;
```
double indent on statement continuations.

trevnorris commented on src/node.cc:
```diff
@@ -2220,6 +2221,35 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+  if (err) return;
+
+  // Get the double array pointer from the Float64Array argument.
+  Local<ArrayBuffer> ab = args[0].As<Float64Array>()->Buffer();
```
Above this do a `CHECK(args[0]->IsFloat64Array())` just above this as a sanity check.

trevnorris replied:
As another sanity check I'd do:
```cpp
Local<Float64Array> array = args[0].As<Float64Array>();
CHECK_EQ(array->Length(), 2);
Local<ArrayBuffer> ab = array->Buffer();
```

trevnorris commented:
@pmuellr Finished initial review. Nice work so far.

pmuellr commented:
Thanks for the review @trevnorris .  All points addressed in the commit I just pushed.

Decided to not go with the "integer" based tests suggested for the previous values (discussion above), using this helper instead, basically just adding an additional test to ensure the value isn't negative (tests added for that):

    function previousValueIsValid(num) {
      if (num == null) return false;
      if (!Number.isFinite(num)) return false;
      if (num < 0) return false;
      return true;
    }

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,53 @@ const assert = process.assert = function(x, msg) {
+    }
+
+    if (!previousValueIsValid(prevValue.system)) {
+      throw new TypeError('value of system property of argument is invalid');
```
checking user input should be the first thing that happens. If it's invalid then there's no point in calling into c++.

pmuellr replied:
yup, will change ...

trevnorris commented on src/node.cc:
```diff
@@ -2220,6 +2221,39 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+
+  // Call libuv to get the values we'll return, and set the return code.
+  int err = uv_getrusage(&rusage);
+  args.GetReturnValue().Set(err);
```
May consider changing this to:
```cpp
args.GetReturnValue().Set(OneByteString(args.GetIsolate(), uv_strerror(err)));
```
and include that message in your error message. would be useful for debugging.

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,53 @@ const assert = process.assert = function(x, msg) {
+    function previousValueIsValid(num) {
+      if (num == null) return false;
+      if (!Number.isFinite(num)) return false;
+      if (num < 0) return false;
```
mind turning this into a single conditional?

pmuellr commented:
just pushed a new commit with changes suggested by @trevnorris last set of comments

trevnorris commented on lib/internal/process.js:
```diff
@@ -22,6 +23,53 @@ const assert = process.assert = function(x, msg) {
+    };
+
+    function previousValueIsValid(num) {
+      return Number.isFinite(num) && num >= 0;
```
Since we're fine with floating point values, in case we opt for nanosecond level precision in the future, I'd prefer the check to be:
```js
return num <= Number.MAX_SAFE_INTEGER && num >= 0;
```
That will grab `NaN`, and verify the value is within proper bounds (since anything over 2^53 can't be a valid value). It will allow string -> number coercion for things like `'0xff'`, but I don't think that's a concern.

pmuellr replied:
I'm happy with that, and adding a comment to explain why the `Number.MAX_SAFE_INTEGER` check is there.

trevnorris commented on src/node.cc:
```diff
@@ -2220,6 +2221,40 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
 }
 
+// Microseconds in a second, used in CPUUsage() below
+#define MICROS_PER_SEC 1000000
```
Could change this to `#define MICROS_PER_SEC 1e6` then below do `MICROS_PER_SEC * rusage.*`. So you can drop the `1.0L`. Just a suggestion, not a blocker.

trevnorris commented on test/parallel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,57 @@
+for (let i = 0; i < 10; i++) {
+  thisUsage = process.cpuUsage();
+  assert(thisUsage.user >= lastUsage.user);
+  assert(thisUsage.system >= lastUsage.system);
```
is it ever possible for `thisUsage.user == lastUsage.user`?

pmuellr replied:
Sure, why not?  Assuming a fast enough processor that can run this code faster than a microsecond.  In fact, the `cpuUsage()` call **with no argument** (returning cpu usage since process start) could in theory return 0's.

Will ensure tests and comments match that assumption.

trevnorris commented on test/parallel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,57 @@
+for (let i = 0; i < 10; i++) {
+  thisUsage = process.cpuUsage(lastUsage);
+  assert(thisUsage.user >= 0);
+  assert(thisUsage.system >= 0);
```
comment doesn't match test. "Ensure that the diffs are greater than 0" != checking `>=`. Should just be `> 0`.

trevnorris commented on test/parallel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,57 @@
+assert.throws(function() { process.cpuUsage({
+  user: 5,
+  system: Number.NEGATIVE_INFINITY
+}); });
```
nit: drop the double `;`. Can compress them to `})});`

pmuellr replied:
jslint (or whatever) **really** wants that extra semi:

```
.../nodejs/node/test/parallel/test-process-cpuUsage.js
  44:3  error  Missing semicolon  semi
  48:3  error  Missing semicolon  semi
```

At least I will change `}); });` to `});});`

trevnorris commented on test/pummel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,26 @@
+
+// Run a busy-loop for 2 seconds.
+const now = Date.now();
+while (Date.now() - now < 2000);
```
Let's drop the time on this to 500ms. 2 seconds is a long time for a test to run. Generally we like tests to run in < 100ms.

pmuellr replied:
Note this is in the "pummel" tests - I copied from [`test/pummel/test-process-hrtime.js`](https://github.com/nodejs/node/blob/23818c6d0ba5503263f97bcd888b1c227b3177af/test/pummel/test-process-hrtime.js#L12), which also uses 2 sec.

Will go ahead and change to 500ms though, don't see a real advantage of using 2 sec over that.

trevnorris commented:
@pmuellr Looking good. Left a few more comments.

pmuellr commented:
pushed a commit with updates per @trevnorris comments

trevnorris commented:
LGTM if CI is happy.

CI: https://ci.nodejs.org/job/node-test-pull-request/2411/

jasnell commented:
Still LGTM. thanks for bearing with us @pmuellr !

pmuellr commented:
Status from the CI build @trevnorris kicked off, had [an error on OS/X](https://ci.nodejs.org/job/node-test-commit/3074/console).  Failure was in [test-tick-processor.js](https://ci.nodejs.org/job/node-test-commit-osx/3102/nodes=osx1010/tapTestReport/test.tap-859/).

In a side chat with Trevor, he indicated this was most likely a flaky.

Happy to look into it more, or if someone wants to kick off another CI build, or can I do that myself?

addaleax commented:
Yeah, that failure is unrelated and is being fixed, no need to worry.

santigimeno commented on doc/api/process.md:
````diff
@@ -504,6 +504,29 @@ the value of `process.config`.*
+argument to the function, to get a diff reading.
+
+```js
+const startUsage  = process.cpuUsage();
````
Extra space before `=`

santigimeno commented on test/parallel/test-process-cpuUsage.js:
```diff
@@ -0,0 +1,64 @@
+// Ensure that the return value is the expected shape.
+function validateResult(result) {
+  assert(result.user != null);
+  assert(result.system != null);
```
Maybe use `assert.notStrictEqual` here.

pmuellr commented:
Thanks for the comments @santigimeno . 

Pushed a new commit.  

Using `assert.notStrictEqual` didn't seem to make sense to me here, as the intent was to check for `null` || `undefined`, so ... would have expected maybe `assert.notEqual`.  But looking at the test as a whole, realized is `Number.isFinite` is already checking null/undefined, and that I wasn't actually checking `result` in the first place, so removed the two field `assert`s, replaced with an `assert(result)` instead.

santigimeno commented:
LGTM. CI: https://ci.nodejs.org/job/node-test-pull-request/2430/

jasnell commented:
CI looks good except for unrelated build bot failure (passed previously on that bot). Landing...

jasnell commented:
Landed in 52cb410. Thanks @pmuellr !

pmuellr commented:
Thanks everyone!  Great way to end the week.
